Source/JavaScriptCore/b3/air/AirOptimizePairedLoadStore.cpp

```diff
@@ -0,0 +1,311 @@
+            case Move32:
+                return ARM64Assembler::isValidSTPImm<32>(offset);
+            case Move:
+                return ARM64Assembler::isValidSTPImm<64>(offset);
+            case MoveDouble:
+                return ARM64Assembler::isValidSTPFPImm<64>(offset);
+            case MoveFloat:
+                return ARM64Assembler::isValidSTPFPImm<32>(offset);
+            default:
+                RELEASE_ASSERT_NOT_REACHED();
+                return false;
+            }
+        };
+
+        int64_t targetOffset = static_cast<int64_t>(target.args[1].offset());
+
+        if (isValidOffset(instOffset) && targetOffset == (instOffset + bytesForWidth(instWidth))) {
+            Inst newInst(pairOpcode, target.origin, inst.args[0], target.args[0], inst.args[1]);
+            logFound(newInst);
+            target = newInst;
+            return true;
+        }
+
+        if (isValidOffset(targetOffset) && (targetOffset + bytesForWidth(instWidth)) == instOffset) {
+            Inst newInst(pairOpcode, target.origin, target.args[0], inst.args[0], target.args[1]);
+            logFound(newInst);
+            target = newInst;
+            return true;
+        }
+
+        // Because str pimm only takes unsigned offset, we tend to pick stackPointerRegister based offsetting.
+        // But it is possible that framePointerRegister based offsetting can offer the benefit here.
+        if (target.args[1].base() == Tmp(CCallHelpers::stackPointerRegister)) {
+            int64_t instOffsetFromFP = instOffset - code.frameSize();
+            int64_t targetOffsetFromFP = targetOffset - code.frameSize();
+
+            if (isValidOffset(instOffsetFromFP) && targetOffsetFromFP == (instOffsetFromFP + bytesForWidth(instWidth))) {
+                Inst newInst(pairOpcode, target.origin, inst.args[0], target.args[0], Arg::addr(Air::Tmp(GPRInfo::callFrameRegister), instOffsetFromFP));
+                logFound(newInst);
+                target = newInst;
+                return true;
+            }
+
+            if (isValidOffset(targetOffsetFromFP) && (targetOffsetFromFP + bytesForWidth(instWidth)) == instOffsetFromFP) {
+                Inst newInst(pairOpcode, target.origin, target.args[0], inst.args[0], Arg::addr(Air::Tmp(GPRInfo::callFrameRegister), targetOffsetFromFP));
+                logFound(newInst);
+                target = newInst;
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+bool optimizePairedLoadStore(Code& code)
+{
+    constexpr bool verbose = false;
+
+    PhaseScope phaseScope(code, "optimizePairedLoadStore");
+
+    if (verbose) {
+        dataLog("Air before an iteration of optimizePairedLoadStore:\n");
+        dataLog(code);
+    }
+
+    bool changed = false;
+    for (BasicBlock* block : code) {
+        unsigned index = block->size();
+        while (index--) {
+            Inst& inst = block->at(index);
+            if (inst.hasNonArgEffects())
+                continue;
```

Comment:
We do not want to merge if it has an effect, for example, Wasm store, which can have a trap.
Then, reordering / merging them changes the semantics since trap and store is observable.

Reply:
Ah I should have read the implementation of `hasNonArgEffects()`, I agree.